bob/src/cluster/support_types.rs

```diff
@@ -57,3 +57,76 @@ impl From<RemoteDeleteError> for Error {
         w.error
     }
 }
+
+pub(crate) struct IndexMap {
+    indexes: Vec<usize>,
+}
+
+impl IndexMap {
+    /// Create empty indexes map
+    pub(crate) fn new() -> Self {
+        Self { indexes: vec![] }
+    }
+
+    /// Create with indexes with `false` values
+    pub(crate) fn where_not_exists(data: &[bool]) -> Self {
+        Self {
+            indexes: data
+                .iter()
+                .enumerate()
+                .filter(|(_, f)| !**f)
+                .map(|(i, _)| i)
+                .collect(),
+        }
+    }
+
+    /// Check that indexes are empty
+    pub(crate) fn is_empty(&self) -> bool {
+        self.indexes.is_empty()
+    }
+
+    /// Get number of indexes in the map
+    pub(crate) fn len(&self) -> usize {
+        self.indexes.len()
+    }
+
+    /// Add index
+    pub(crate) fn push(&mut self, item: usize) {
+        debug_assert!(!self.indexes.contains(&item));
+        self.indexes.push(item);
+    }
+
+    /// Collect data with allowed indexes
+    pub(crate) fn collect<T: Clone>(
+        &self,
+        data: &[T]
+    ) -> Vec<T> {
+        let mut result = Vec::with_capacity(self.indexes.len());
+        for &i in &self.indexes {
+            if i < data.len() {
+                result.push(data[i].clone());
+            }
+        }
+        result
+    }
+
+    /// Update elements in original according to internal indexes mapping
+    pub(crate) fn update_existence(&self, original: &mut [bool], mapped: &[bool]) {
+        let max = original.len();
+        let len = std::cmp::max(self.indexes.len(), mapped.len());
```

Comment:
Same here. We expect, that `self.indexes.len() == mapped.len()`. This should be enforced with `assert!(self.indexes.len() == mapped.len());`. Here `assert` is better than `debug_assert`, because it allow to detect problem in the production build before entering the loop (in previous case in production the code would still fail on index-of-range)